### PR TITLE
Thoroughly fix the error handling

### PR DESCRIPTION
**Pull request by kw217**
This is a thorough root-and-branch replacement of all the error handling with something sane.

There is now a single error type defined in errors.rs, which has variants for plain Cassandra errors (just a return code), detailed errors (with additional info), internal errors (currently only "unexpected type") or various foreign errors (in particular, problems converting strings to/from C).

This error type is used appropriately everywhere, and has appropriate conversions as necessary.

In addition I've audited and removed all (well, almost all) uses of `unwrap`, `expect`, `panic!`, `unimplemented`, and friends, replacing them with principled use of `?` and `Result`. There are a few that remain where they are appropriate, and one (Statement::new) I'm not 100% happy with but think it best to leave.

This is very much a breaking API change - several return types have been changed (from `T` to `Result<T, Error>`, or from `Result<T, WrongError>` to `Result<T, Error>`, usually, though some other changes have been made too).

Some other choice details:
* Enhance the `enhance_nullary_enum` macro to cope with omitted variants. We don't want `CASS_OK` or `CASS_ERROR_LAST_ENTRY` to count as errors which you need to consider in `match`; we should ensure these just never happen.
* Move some error-handling code from `futures` to `error` module. Add some other nice error building helpers. Remove direct tests of `CASS_OK` elsewhere and depend on helpers in `error` module.
* Define WriteType and ValueType as nice modern enums.
* Fix cassandra_sys in https://github.com/Metaswitch/cassandra-sys-rs/pull/7 to not attempt to add limited error-chain support to the bare return code; instead use an extension trait in this crate to do it (adding `to_result`). Also add some missing symbols while we're at it.
* Get rid of all uses of `.chain_err(|| "pointless text")`; the Cassandra errors speak for themselves and don't need the addition of an extra wrapper.

Enjoy!

### CODE REVIEW

**Comment by kw217** [comment]
(now rebased with correct signed-off-by lines!)

File: src/cassandra/column.rs [hunk 1]
```diff
@@ -124,22 +119,37 @@ impl Debug for Column {
             CASS_VALUE_TYPE_TIMEUUID => write!(f, "TIMEUUID Cassandra type"),
             CASS_VALUE_TYPE_INET => write!(f, "INET Cassandra type"),
             CASS_VALUE_TYPE_LIST => {
-                for item in self.set_iter().expect("a list should always be able to return a set iterator") {
-                    write!(f, "LIST {:?}", item)?
+                match self.set_iter() {
```
Comment (kw217) [inline src/cassandra/column.rs]: Don't worry too much about these - I make a better fix to them in the next PR, https://github.com/Metaswitch/cassandra-rs/pull/23

File: src/cassandra/result.rs [hunk 2]
```diff
@@ -85,13 +82,16 @@ impl CassResult {
...
     /// Gets the column name at index for the specified result.
-    pub fn column_name(&self, index: usize) -> String {
+    pub fn column_name(&self, index: usize) -> Result<&str> {
         unsafe {
             let name = mem::zeroed();
             let name_length = mem::zeroed();
-            cass_result_column_name(self.0, index, name, name_length);
-            let slice = slice::from_raw_parts(name as *const u8, name_length as usize);
-            str::from_utf8(slice).expect("must be utf8").to_owned()
+            cass_result_column_name(self.0, index, name, name_length).to_result(())
```
Comment (kw217) [inline src/cassandra/result.rs]: Spot the bug - fixed in https://github.com/Metaswitch/cassandra-rs/pull/23

**Comment by kw217** [comment]
Note to self - fix Cargo.toml to point to the new 0.10 library once it's merged.

File: src/cassandra/result.rs (hunk 2 above)
Comment (angusi) [inline src/cassandra/result.rs]: Surely the real solution is to move everything into Rust and stop using Unsafe code? 😈

**Comment by angusi** [comment]
Oh - do we also need a version bump, if we've an API breaking change?

**Comment by kw217** [comment]
I've not yet released the previous breaking change. I'm holding off so I can release them all together, as v0.9.0. (hence this is -pre1). I guess technically I could have bumped to -pre2? But it doesn't really matter at this stage, especially since versions are ignored by Cargo on git dependencies.

**Comment by kw217** [comment]
Thanks Angus.

TODO: depend on updated cassandra-sys crate, then merge.

**Pull request merged**